exercise-book/src/files-match-result.md

```diff
@@ -0,0 +1,277 @@
+
+✅ Check the documentation for the exact type
+    [File::open](https://doc.rust-lang.org/std/fs/struct.File.html#method.open)
+    returns.
+
+✅ Handle the `Result` using `match` to get to the inner type. Link the two possible patterns, `Ok(file)` and `Err(e)` to an an appropriate expression, for example: `println!("File opened")` and `panic!("Problem opening the file: {:?}", e)`.
+
+✅ Fix the location of the file so that the program no longer panics. 
+
+<details>
+  <summary>Click me</summary>
+
+```rust, ignore
+{{#include ../../exercise-solutions/files-match-result/examples/step_1.rs}}
+```
+
+</details>
+
+TIP: IDEs often provide a "quick fix" to roll out all match arms quickly
+
+### Step 2: Reading the File content to a String and Error propagation.
+
+✅ Import `std::io::prelude::*`
+
+Take a look at [Read::read\_to\_string](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_string). The method takes in a mutable empty `String`, and writes the content of a file to this buffer. The method returns a `Result<usize, Error>`, where the usize is the number of bytes that have been written to the buffer. Handling this Result, will not yield the `String` of file content. For a simple program, handling it with an 
+`.unwrap()` would be sufficient, but for bigger code bases this is not helpful, so we want to propagate the Error.
+
+✅ Add `Result<(), Error>` as return type to `fn main()` and `Ok(())` as the last line of `fn main()`.
+  
+
+✅ Create an empty `String` that serves as buffer and bind it to a mutable variable. 
+
+✅ Call the `.read_to_string()` method on the `File` object. The method takes in the `String` buffer and is followed by the `?` operator. If the method returns an `Error` it is propagated to the instance that called `fn main()`. If the method returns the `Ok` value, the program proceeds as planned.
+
+✅ Use `println!` to print the content of the `String` buffer.
+
+<details>
+  <summary>Click me</summary>
+
+```rust, ignore
+{{#include ../../exercise-solutions/files-match-result/examples/step_2.rs}}
+```
+</details>
+
+### Task 3: Read the lines into a `BufReader` and count them!
+
+✅ Add the following imports:
+
+```rust
+use std::io::{ BufReader, BufRead,};
+```
+
+✅ Take a look at the documentation of [BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html). BufReader is a struct that adds buffering to any reader. It implements the
+[`BufRead`](https://doc.rust-lang.org/std/io/trait.BufRead.html#) trait. In short this means, that methods that are defined for `BufRead` can be used for `BufReader`. For example the [`lines()`](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines) method.
+
+✅ Construct a `BufReader` around the file.
+
+✅ The `lines()`- method returns an Iterator over the file’s lines. Iterate over the lines with a for loop to count them.
+
+✅ Print the number of lines the file contains.
+
+✅ You don’t have to handle the `Result` that is returned from `.lines()`, why?
```

Comment:
`.lines()` doesn't return a `Result`?

Reply:
I deleted the question as that is kept better for the discussion afterward

Reply:
Ah, the `Item` you get out of the Iterator is a `Result`. Maybe that's what was meant here. Maybe...

```suggestion
✅ When you go through the for loop on `.lines()`, each time around the loop you get  a `std::io::Result<&str>` and not a string-slice (`&str`). Why? Where does the `BufReader` get its input from?
```

Reply:
OK!